tap-version-14-specification.md

```diff
@@ -0,0 +1,741 @@
+the text after # SKIP\S*\s+ as a reason for skipping.
+
+```
+ok 23 # skip Insufficient flogiston pressure.
+```
+
+Similarly, one can include an explanation in a Plan line, emitted if
+the test file is skipped completely:
+
+```
+1..0 # SKIP WWW::Mechanize not installed
+```
+
+## SUBTESTS
+
+A Subtest contains a block of test points grouped together.  This is
+useful for the output of test runners that collect several child
+processes' TAP streams, and output the result in TAP for another
+harness to interpret, or for tests with a predictable number of
+blocks, but where each block might not have a predictable number of
+tests.
+
+A Subtest is a TAP14 stream indented 4 spaces.  It has the following
+format:
+
+- A comment line with the name of the subtest: `/^ {4}# Subtest: (.*)$/`
+- A TAP14 stream, indented 4 spaces.
+- A (non-indented) test point indicating the success or failure of the
+  subtest, with a description matching the Subtest's name.
+
+For example:
+
+```
+TAP version 14
+1..2
+ok 1 - About to do a subtest
+    # Subtest: this is the subtest
+    ok 1 - This is ok
+    ok 2 - I am ok with how things are proceeding
+    not ok 3 - infinite loops # TODO Solve halting problem
+    not ok 4 - this is just a failure
+    1..4
+not ok 2 - this is the subtest
+  ---
+  plan: 4
+  count: 4
+  passing: 2
+  todo: 1
+  failed: 1
+  exitCode: 1
+  ...
+```
+
+If not bailed out, a Subtest MUST be introduced by the indented
+`# Subtest: ...` comment, and MUST be followed with a test point whose
+description matches the Subtest name.  If the Subtest is not
+successful, then the Subtest test point MUST be a `not ok` test point.
+
+A `Bail out!` in a Subtest SHOULD be immediately followed by a
+`Bail out!` in the parent test.
+
+If a Subtest TAP stream does not include a version number, it MUST be
+interpreted as a TAP14 stream anyway.  This is a backwards-compatible
+affordance for pre-TAP14 parsers that complain about seeing a `TAP
+version` line multiple times within a TAP stream.
+
+## OTHER LINES
+
+### Bail out!
+
+As an emergency measure a test script can decide that further tests
+are useless (e.g. missing dependencies) and testing should stop
+immediately. In that case the test script prints the magic words
+
+```
+Bail out!
+```
+
+to standard output. Any message after these words must be displayed by
+the interpreter as the reason why testing must be stopped, as in
+
+```
+Bail out! MySQL is not running.
+```
+
+A Harness SHOULD ignore all lines that come after a bail out.
+
+### Diagnostics
+
+Additional information may be put into the testing output on separate
+lines. Diagnostic lines should begin with a #, which the harness must
+ignore, at least as far as analyzing the test results. The harness is
+free, however, to display the diagnostics.
+
+YAML blocks are intended to replace diagnostics for most purposes but
+consumers should maintain backwards compatibility by supporting them.
+
+### Pragma
```

Comment:
Seems ok to me.  It's a way for a producer to say, within a given context, that deviations from the "approved" syntax should be treated as syntax errors.

It's different from JS's `'use strict'` pragma since that actually opts into a few profound semantic changes, but I think there's a valid use case for "this test should not dump any junk to stdout, and if it does, then that's an error".

There's 2 implementations now, since I added support to the JS `tap-parser` so that it would handle `Test::Harness`'s tests :)